### PR TITLE
add db infrastructure

### PR DESCRIPTION
**Pull request by Melchalk**

### CODE REVIEW

**Comment by SolovyovDmitry** [comment]
image
this stuff must be filled in

File: src/ScheduleService.Models.Db/DbCategory.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace LT.DigitalOffice.ScheduleService.Models.Db;
+
+public class DbCategory
+{
+  public const string TableName = "Categories";
+
+  public Guid Id { get; set; }
+
+  [MaxLength(20)]
```
Comment (SolovyovDmitry) [inline src/ScheduleService.Models.Db/DbCategory.cs]: It's not very useful. Constraints like this comes from migration

File: src/ScheduleService.Models.Db/DbTask.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  public string Description { get; set; }
+
+  public int Type { get; set; }
+
+  public Guid GroupId { get; set; }
+  public Guid WorkspaceId { get; set; }
+
+  public DbGroup Group { get; set; }
+  public DbWorkspace Workspace { get; set; }
+  public IList<DbCategory> Categories { get; set; } = new List<DbCategory>();
```
Comment (SolovyovDmitry) [inline src/ScheduleService.Models.Db/DbTask.cs]: split and order like this:

public const string TableName = "Tasks";

public Guid Id { get; set; }
public Guid GroupId { get; set; }
public Guid WorkspaceId { get; set; }
public string Name { get; set; }
public string Description { get; set; }
public int Type { get; set; }

public DbGroup Group { get; set; }
public DbWorkspace Workspace { get; set; }
public IList<DbCategory> Categories { get; set; } = new List<DbCategory>();

Comment (SolovyovDmitry) [inline src/ScheduleService.Models.Db/DbTask.cs]: can it be just new()?

**Review by SolovyovDmitry — CHANGES_REQUESTED** [review]
Models are OK. Add migration to this PR

File: src/ScheduleService.Models.Db/DbWorkspace.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  public void Configure(EntityTypeBuilder<DbWorkspace> builder)
+  {
+    builder
+      .ToTable(DbWorkspace.TableName);
+
+    builder
+      .HasKey(a => a.Id);
+
+    builder
+      .HasMany(u => u.Tasks)
```
Comment (VLiamin) [inline src/ScheduleService.Models.Db/DbWorkspace.cs]: What does u mean? May be use w(Workspace)?

File: src/ScheduleService.Models.Db/DbTask.cs (hunk 2 above)
Comment (Melchalk) [reply src/ScheduleService.Models.Db/DbTask.cs]: I have questions
Will restrictions on properties be set in validation then? Don't you need to add them in the attributes? And then configure the restrictions in the database in the migration file?

Comment (VLiamin) [reply src/ScheduleService.Models.Db/DbTask.cs]: Such restrictions are useful in requests coming from the front. There they will actually check the data. Here it’s too late.

File: src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs [hunk 4]
```diff
@@ -0,0 +1,142 @@
+﻿using LT.DigitalOffice.ScheduleService.Models.Db;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Migrations;
+using System;
+
+namespace LT.DigitalOffice.ScheduleService.Data.Provider.MsSql.Ef.Migrations;
+
+[DbContext(typeof(ScheduleServiceDbContext))]
+[Migration("20231217210755_InitialCreate")]
+
```
Comment (VLiamin) [inline src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs]: empty line

File: src/ScheduleService.Models.Db/DbTask.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+public class DbTask
+{
+  public const string TableName = "Tasks";
+
+  public Guid Id { get; set; }
+  public string Name { get; set; }
+  public string Description { get; set; }
+  public int Type { get; set; } //enum
+  public Guid GroupId { get; set; }
+  public Guid WorkspaceId { get; set; }
```
Comment (VLiamin) [inline src/ScheduleService.Models.Db/DbTask.cs:14]: @SolovyovDmitry is it normal that tables do not contain fields(CreatedBy, ModifiedBy ...)?

File: src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      name: DbGroup.TableName,
+      columns: table => new
+      {
+        Id = table.Column<Guid>(nullable: false),
+        Name = table.Column<string>(nullable: false, maxLength: 50),
+      },
+      constraints: table =>
+      {
+        table.PrimaryKey($"PK_{DbGroup.TableName}", x => x.Id);
+        //table.UniqueConstraint($"UX_{DbGroup.TableName}_Name_unique", x => x.Name);
```
Comment (SolovyovDmitry) [inline src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs]: why not?

File: src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+  private void CreateTableCategories(MigrationBuilder migrationBuilder)
+  {
+    migrationBuilder.CreateTable(
+      name: DbCategory.TableName,
+      columns: table => new
+      {
+        Id = table.Column<Guid>(nullable: false),
+        Name = table.Column<string>(nullable: false, maxLength: 20),
+        Color = table.Column<string>(nullable: false, maxLength: 50),
```
Comment (SolovyovDmitry) [inline src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs:68]: What's the idea for color?

File: src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    migrationBuilder.CreateTable(
+      name: DbTask.TableName,
+      columns: table => new
+      {
+        Id = table.Column<Guid>(nullable: false),
+        Name = table.Column<string>(nullable: false, maxLength: 20),
+        Description = table.Column<string>(nullable: false, maxLength: 300),
+        Type = table.Column<int>(nullable: false),
+        GroupId = table.Column<Guid>(nullable: false),
+        WorkspaceId = table.Column<Guid>(nullable: false),
```
Comment (SolovyovDmitry) [inline src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs]: Move it right after id

File: src/ScheduleService.Models.Db/DbGroup.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace LT.DigitalOffice.ScheduleService.Models.Db;
+
+public class DbGroup
+{
+  public const string TableName = "Groups";
+
+  public Guid Id { get; set; }
+  public string Name { get; set; }
+  public Guid WorkspaceId { get; set; }
```
Comment (SolovyovDmitry) [inline src/ScheduleService.Models.Db/DbGroup.cs:13]: same

File: src/ScheduleService.Models.Db/DbTask.cs (hunk 5 above)
Comment (SolovyovDmitry) [reply src/ScheduleService.Models.Db/DbTask.cs:14]: Same about order

Comment (SolovyovDmitry) [reply src/ScheduleService.Models.Db/DbTask.cs:14]: As @VLiamin said, you need to add CreatedBy,CreatedAtUtc columns for all entities; ModifiedBy, ModifiedAtUtc for entities that can be edited; IsActive for entities that can be archived(soft deleted).!!! Except intermediate many-to-many tables

File: src/ScheduleService.Models.Db/DbGroup.cs (hunk 9 above)
Comment (Melchalk) [reply src/ScheduleService.Models.Db/DbGroup.cs:13]: you mean like this?

  public Guid Id { get; set; }
  public Guid WorkspaceId { get; set; }
  public string Name { get; set; }

File: src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs (hunk 6 above)
Comment (Melchalk) [reply src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs]: I forgot to ask about it
Can the different group/workspace and another have the same name or something like this?
And do we need to nave this rule in db or it is need to check earlier

File: src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs (hunk 7 above)
Comment (Melchalk) [reply src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs:68]: We think about dictionary with key - name of color and value - code if color

File: src/ScheduleService.Models.Db/DbTask.cs (hunk 5 above)
Comment (Melchalk) [reply src/ScheduleService.Models.Db/DbTask.cs:14]: Do i need to just add it as other properties or with additional logic? And there are only five of them?

File: src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs (hunk 6 above)
Comment (Melchalk) [reply src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs]: +

File: src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs (hunk 8 above)
Comment (Melchalk) [reply src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs]: +

File: src/ScheduleService.Models.Db/DbGroup.cs (hunk 9 above)
Comment (Melchalk) [reply src/ScheduleService.Models.Db/DbGroup.cs:13]: +

File: src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs (hunk 6 above)
Comment (Melchalk) [reply src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs:26]: +

File: src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs (hunk 7 above)
Comment (Melchalk) [reply src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs:68]: +

File: src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs (hunk 8 above)
Comment (Melchalk) [reply src/ScheduleService.Data.Provider.MsSql.Ef/Migrations/20231217210755_InitialCreate.cs:87]: +

File: src/ScheduleService.Models.Db/DbGroup.cs (hunk 9 above)
Comment (Melchalk) [reply src/ScheduleService.Models.Db/DbGroup.cs:14]: +

File: src/ScheduleService.Models.Db/DbTask.cs (hunk 5 above)
Comment (Melchalk) [reply src/ScheduleService.Models.Db/DbTask.cs:14]: +

File: src/ScheduleService.Models.Db/DbCategoriesTask.cs [hunk 10]
```diff
@@ -10,6 +10,8 @@ public class DbCategoriesTask
 
   public Guid TaskId { get; set; }
   public Guid CategoryId { get; set; }
+  public Guid CreatedBy { get; set; }
+  public DateTime CreatedAtUtc { get; set; }
```
Comment (SolovyovDmitry) [inline src/ScheduleService.Models.Db/DbCategoriesTask.cs]: m-m entities doesn't have to include this

**Review by SolovyovDmitry — APPROVED** [review]

**Review by VLiamin — APPROVED** [review]

**Pull request merged**